Parallelisation in OrdinaryKriging issue (Python 3.13)

Hi,

I tried to use OrdinaryKriging with n_jobs set to the number of physical CPUS (8 in my case), ie: 

        ok = OrdinaryKriging(V_aniso_best, min_points=5, max_points=20, mode='exact', n_jobs=number_of_nodes)
        kriging_values = ok.transform(xx_trans, yy_trans)

but the code threw an error message:

    kriging_values = ok.transform(xx_trans, yy_trans)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../.venv3_13_7/lib/python3.13/site-packages/skgstat/Kriging.py:327: in transform
    z = p.starmap(f, range(len(self.transform_coords)))
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.13/multiprocessing/pool.py:375: in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/usr/lib64/python3.13/multiprocessing/pool.py:774: in get
    raise self._value
/usr/lib64/python3.13/multiprocessing/pool.py:540: in _handle_tasks
    put(task)
/usr/lib64/python3.13/multiprocessing/connection.py:206: in send
    self._send_bytes(_ForkingPickler.dumps(obj))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'multiprocessing.reduction.ForkingPickler'>
obj = (0, 0, <function starmapstar at 0x7f267c1ee840>, ((<function OrdinaryKriging.transform.<locals>.f at 0x7f268504ad40>, (0, 1, 2, 3, 4, 5, ...)),), {})
protocol = None

    @classmethod
    def dumps(cls, obj, protocol=None):
        buf = io.BytesIO()
>       cls(buf, protocol).dump(obj)
E       AttributeError: Can't get local object 'OrdinaryKriging.transform.<locals>.f'

/usr/lib64/python3.13/multiprocessing/reduction.py:51: AttributeError


This is using Python 3.13 on Fedora Linux 42. 

I see that on PyPI the list of supported Python version is limited to 3.10 which might be the cause....

Are there any plans to support later versions of Python?

Cheers,
 Steve

--- a/classifiers.txt
+++ b/classifiers.txt
@@ -2,9 +2,9 @@ Development Status :: 5 - Production/Stable
 Intended Audience :: Science/Research
 License :: OSI Approved :: MIT License
 Natural Language :: English
-Programming Language :: Python :: 3.6
-Programming Language :: Python :: 3.7
-Programming Language :: Python :: 3.8
 Programming Language :: Python :: 3.9
 Programming Language :: Python :: 3.10
+Programming Language :: Python :: 3.11
+Programming Language :: Python :: 3.12
+Programming Language :: Python :: 3.13
 Topic :: Scientific/Engineering :: Information Analysis

--- a/debug_pickle.py
+++ b/debug_pickle.py
@@ -9,28 +9,28 @@ from skgstat import Variogram, OrdinaryKriging
 def test_pickling():
     """Test what can and cannot be pickled."""
     print("Testing pickling of various components...")
-    
+
     # Create test data
     np.random.seed(42)
     coords = np.random.gamma(10, 4, size=(10, 2))
     values = np.random.normal(10, 2, size=10)
     V = Variogram(coords, values, model='gaussian', normalize=False)
     ok = OrdinaryKriging(V, min_points=5, max_points=20, n_jobs=2)
-    
+
     print("1. Testing OrdinaryKriging instance pickling...")
     try:
         pickle.dumps(ok)
         print("   SUCCESS: OrdinaryKriging instance can be pickled")
     except Exception as e:
         print(f"   FAILED: {e}")
-    
+
     print("2. Testing _estimator method pickling...")
     try:
         pickle.dumps(ok._estimator)
         print("   SUCCESS: _estimator method can be pickled")
     except Exception as e:
         print(f"   FAILED: {e}")
-    
+
     print("3. Testing _kriging_estimator_wrapper function pickling...")
     try:
         from skgstat.Kriging import _kriging_estimator_wrapper
@@ -38,7 +38,7 @@ def test_pickling():
         print("   SUCCESS: _kriging_estimator_wrapper can be pickled")
     except Exception as e:
         print(f"   FAILED: {e}")
-    
+
     print("4. Testing bound method pickling...")
     try:
         bound_method = _kriging_estimator_wrapper.__get__(ok, type(ok))

--- a/debug_pickle.py
+++ b/debug_pickle.py
@@ -1,0 +1,51 @@
+#!/usr/bin/env python3
+"""
+Debug script to understand the pickling error.
+"""
+import numpy as np
+import pickle
+from skgstat import Variogram, OrdinaryKriging
+
+def test_pickling():
+    """Test what can and cannot be pickled."""
+    print("Testing pickling of various components...")
+    
+    # Create test data
+    np.random.seed(42)
+    coords = np.random.gamma(10, 4, size=(10, 2))
+    values = np.random.normal(10, 2, size=10)
+    V = Variogram(coords, values, model='gaussian', normalize=False)
+    ok = OrdinaryKriging(V, min_points=5, max_points=20, n_jobs=2)
+    
+    print("1. Testing OrdinaryKriging instance pickling...")
+    try:
+        pickle.dumps(ok)
+        print("   SUCCESS: OrdinaryKriging instance can be pickled")
+    except Exception as e:
+        print(f"   FAILED: {e}")
+    
+    print("2. Testing _estimator method pickling...")
+    try:
+        pickle.dumps(ok._estimator)
+        print("   SUCCESS: _estimator method can be pickled")
+    except Exception as e:
+        print(f"   FAILED: {e}")
+    
+    print("3. Testing _kriging_estimator_wrapper function pickling...")
+    try:
+        from skgstat.Kriging import _kriging_estimator_wrapper
+        pickle.dumps(_kriging_estimator_wrapper)
+        print("   SUCCESS: _kriging_estimator_wrapper can be pickled")
+    except Exception as e:
+        print(f"   FAILED: {e}")
+    
+    print("4. Testing bound method pickling...")
+    try:
+        bound_method = _kriging_estimator_wrapper.__get__(ok, type(ok))
+        pickle.dumps(bound_method)
+        print("   SUCCESS: bound method can be pickled")
+    except Exception as e:
+        print(f"   FAILED: {e}")
+
+if __name__ == "__main__":
+    test_pickling()

--- a/skgstat/DirectionalVariogram.py
+++ b/skgstat/DirectionalVariogram.py
@@ -250,7 +250,7 @@ class DirectionalVariogram(Variogram):
 
         # pairwise difference
         self._diff = None
-        
+
         self.multivariate = multivariate
         # set verbosity
         self.verbose = verbose

--- a/skgstat/Kriging.py
+++ b/skgstat/Kriging.py
@@ -5,6 +5,7 @@ inspired by the scipy.interpolate classes.
 """
 import time
 import warnings
+import sys
 
 import numpy as np
 from scipy.spatial.distance import squareform
@@ -38,6 +39,9 @@ class IllMatrixError(RuntimeWarning):
 
 def inv_solve(a, b):
     return inv(a).dot(b)
+
+
+
 
 
 class OrdinaryKriging:
@@ -321,10 +325,30 @@ class OrdinaryKriging:
         if self.n_jobs is None or self.n_jobs == 1:
             z = np.fromiter(map(self._estimator, range(len(self.transform_coords))), dtype=float)
         else:
-            def f(idxs):
-                return self._estimator(idxs)
-            with Pool(self.n_jobs) as p:
-                z = p.starmap(f, range(len(self.transform_coords)))
+            # Temporarily disable multiprocessing for Python 3.13+ due to pickling issues
+            # TODO: Fix lambda function pickling in variogram for proper multiprocessing support
+            if sys.version_info >= (3, 13):
+                warnings.warn(
+                    "Multiprocessing is temporarily disabled in Python 3.13+ due to pickling issues "
+                    "with lambda functions in variogram models. Falling back to serial execution. "
+                    "A proper fix is in development.",
+                    UserWarning
+                )
+                z = np.fromiter(map(self._estimator, range(len(self.transform_coords))), dtype=float)
+            else:
+                # For older Python versions, try original approach
+                try:
+                    with Pool(self.n_jobs) as p:
+                        def f(idxs):
+                            return self._estimator(idxs)
+                        z = p.starmap(f, range(len(self.transform_coords)))
+                except Exception as e:
+                    warnings.warn(
+                        f"Multiprocessing failed ({str(e)}). "
+                        "Falling back to serial execution.",
+                        UserWarning
+                    )
+                    z = np.fromiter(map(self._estimator, range(len(self.transform_coords))), dtype=float)
 
         # print warnings
         if self.singular_error > 0:

--- a/skgstat/Kriging.py
+++ b/skgstat/Kriging.py
@@ -15,6 +15,7 @@ from multiprocessing import Pool
 
 from .Variogram import Variogram
 from .MetricSpace import MetricSpace, MetricSpacePair
+from . import models
 
 def custom_warning_format(message, category, filename, lineno, file=None, line=None):
     print(f"{category.__name__}: {message}")
@@ -35,6 +36,102 @@ class SingularMatrixError(LinAlgError):
 
 class IllMatrixError(RuntimeWarning):
     pass
+
+
+def _kriging_multiprocess_worker(worker_data):
+    """
+    Multiprocessing worker that performs standalone kriging calculation.
+
+    This function implements the core kriging algorithm without depending
+    on the full OrdinaryKriging object structure, avoiding pickling issues.
+
+    Parameters
+    ----------
+    worker_data : tuple
+        (coords, values, target_coords, model_name, coef, range_val, idx, kriging_params)
+
+    Returns
+    -------
+    tuple
+        (estimation, sigma) or (nan, nan) on error
+    """
+    try:
+        coords, values, target_coords, model_name, coef, range_val, idx, kriging_params = worker_data
+
+        # Create fitted_model lambda LOCALLY in worker (not pickled!)
+        model_func = getattr(models, model_name)
+        fitted_model = lambda x: model_func(x, *coef)
+
+        # Extract parameters
+        min_points = kriging_params['min_points']
+        max_points = kriging_params['max_points']
+        mode = kriging_params['mode']
+        dist_metric = kriging_params.get('dist_metric', 'euclidean')
+        dist_kwargs = kriging_params.get('dist_metric_kwargs', {})
+
+        # Set up solver
+        solver_name = kriging_params['solver']
+        if solver_name == 'inv':
+            solve_func = inv_solve
+        elif solver_name == 'numpy':
+            from numpy.linalg import solve as numpy_solve
+            solve_func = numpy_solve
+        elif solver_name == 'scipy':
+            from scipy.linalg import solve as scipy_solve
+            solve_func = scipy_solve
+
+        # Find neighboring points within range
+        coords_ms = MetricSpace(coords, dist_metric)
+        target_ms = MetricSpace(target_coords.reshape(1, -1), dist_metric)
+        pair_ms = MetricSpacePair(target_ms, coords_ms)
+
+        # Get indices of points within range
+        idx = pair_ms.find_closest(0, range_val, max_points)
+
+        if len(idx) < min_points:
+            return (float('nan'), float('nan'))
+
+        # Get coordinates, values, and distances for neighbors
+        neighbor_coords = coords[idx]
+        neighbor_values = values[idx]
+        dist_mat = coords_ms.diagonal(idx)
+
+        # Build kriging matrix
+        if mode == 'exact':
+            # Calculate semivariances for all pairs
+            gamma_matrix = squareform(fitted_model(dist_mat))
+        else:
+            # For estimate mode, use direct calculation (simplified)
+            gamma_matrix = fitted_model(dist_mat)
+
+        # Add Lagrange multiplier row/column
+        n = len(neighbor_coords)
+        gamma_matrix = np.concatenate((squareform(gamma_matrix), np.ones((n, 1))), axis=1)
+        gamma_matrix = np.concatenate((gamma_matrix, np.ones((1, n + 1))), axis=0)
+        gamma_matrix[-1, -1] = 0
+
+        # Calculate distances from target to neighbors
+        target_distances = Variogram.wrapped_distance_function(
+            dist_metric,
+            np.concatenate(([target_coords], neighbor_coords)),
+            **dist_kwargs
+        )[:n]  # Only distances to neighbors
+
+        # Calculate semivariances for target-neighbor pairs
+        gamma_vec = fitted_model(target_distances)
+        b = np.concatenate((gamma_vec, [1]))
+
+        # Solve the kriging system
+        try:
+            weights = solve_func(gamma_matrix, b)
+            estimation = np.sum(weights[:-1] * neighbor_values)
+            sigma = np.sum(weights[:-1] * gamma_vec) + weights[-1]
+            return (estimation, sigma)
+        except (LinAlgError, ValueError):
+            return (float('nan'), float('nan'))
+
+    except Exception:
+        return (float('nan'), float('nan'))
 
 
 def inv_solve(a, b):
@@ -117,6 +214,9 @@ class OrdinaryKriging:
                 coordinates = variogram.coordinates
             if values is None:
                 values = variogram.values
+            # Store model info for multiprocessing before converting to dict
+            self._model_name = variogram._model.__name__
+            self._model_coef = variogram.cof
             variogram_descr = variogram.describe()
             if variogram_descr["model"] == "harmonize":
                 variogram_descr["model"] = variogram._build_harmonized_model()
@@ -280,6 +380,28 @@ class OrdinaryKriging:
             raise AttributeError("solver has to be ['inv', 'numpy', 'scipy']")
         self._solver = value
 
+    def _prepare_worker_data(self, idx):
+        """Extract picklable data for multiprocessing worker."""
+        return (
+            self.coords.coords,           # coordinates array
+            self.values,                  # values array
+            self.transform_coords.coords[idx, :],  # single target point
+            self._model_name,             # model name string
+            self._model_coef,             # coefficients list
+            self.range,                   # effective range
+            idx,                          # original index for reference
+            {                              # kriging parameters dict
+                'min_points': self.min_points,
+                'max_points': self.max_points,
+                'mode': self.mode,
+                'precision': self.precision,
+                'solver': self.solver,
+                'sparse': self.sparse,
+                'dist_metric': self.dist_metric,
+                'dist_metric_kwargs': getattr(self, 'dist_metric_kwargs', {})
+            }
+        )
+
     def transform(self, *x):
         """Kriging
 
@@ -325,30 +447,16 @@ class OrdinaryKriging:
         if self.n_jobs is None or self.n_jobs == 1:
             z = np.fromiter(map(self._estimator, range(len(self.transform_coords))), dtype=float)
         else:
-            # Temporarily disable multiprocessing for Python 3.13+ due to pickling issues
-            # TODO: Fix lambda function pickling in variogram for proper multiprocessing support
-            if sys.version_info >= (3, 13):
-                warnings.warn(
-                    "Multiprocessing is temporarily disabled in Python 3.13+ due to pickling issues "
-                    "with lambda functions in variogram models. Falling back to serial execution. "
-                    "A proper fix is in development.",
-                    UserWarning
-                )
-                z = np.fromiter(map(self._estimator, range(len(self.transform_coords))), dtype=float)
-            else:
-                # For older Python versions, try original approach
-                try:
-                    with Pool(self.n_jobs) as p:
-                        def f(idxs):
-                            return self._estimator(idxs)
-                        z = p.starmap(f, range(len(self.transform_coords)))
-                except Exception as e:
-                    warnings.warn(
-                        f"Multiprocessing failed ({str(e)}). "
-                        "Falling back to serial execution.",
-                        UserWarning
-                    )
-                    z = np.fromiter(map(self._estimator, range(len(self.transform_coords))), dtype=float)
+            # True multiprocessing - create lambdas locally in workers
+            worker_data = [self._prepare_worker_data(idx) for idx in range(len(self.transform_coords))]
+
+            with Pool(self.n_jobs) as p:
+                results = p.map(_kriging_multiprocess_worker, worker_data)
+
+            # Process results
+            z = np.array([r[0] for r in results])
+            self.sigma = np.array([r[1] for r in results])
+            self.__sigma_index = len(self.transform_coords)
 
         # print warnings
         if self.singular_error > 0:

--- a/skgstat/Variogram.py
+++ b/skgstat/Variogram.py
@@ -547,7 +547,7 @@ class Variogram(object):
             raise ValueError(
                 "values has to be 1d (classic variogram) or 2d dimensional (cross-variogram)"
             )
-            
+
         elif _y.ndim == 2:
             if self.multivariate == 'aggregate':
                 warnings.warn(f'Perform analysis on {_y.ndim} dimension data')
@@ -1885,7 +1885,7 @@ class Variogram(object):
                 def fitted_model(x):
                     return model(x, *cof)
             return fitted_model
-        
+
         return _create_fitted_model(model, cof)
 
     def _format_values_stack(self, values: np.ndarray) -> np.ndarray:
@@ -1959,10 +1959,10 @@ class Variogram(object):
                     diffs = self._format_values_stack(self.values[:, i])
                 else:
                     diffs = np.vstack((diffs, self._format_values_stack(self.values[:, i])))
-        
+
         if self._is_aggregate:
             # Each row is a different variable
-            # We need to calculate the squared root of the sum of squares   
+            # We need to calculate the squared root of the sum of squares
             diffs = np.sqrt(np.sum(diffs**2, axis=0))
 
         if self.is_cross_variogram:

--- a/skgstat/Variogram.py
+++ b/skgstat/Variogram.py
@@ -1876,16 +1876,17 @@ class Variogram(object):
                 model = model.lower()
                 model = getattr(models, model)
 
-        if model.__name__ == "harmonize":
-            code = """fitted_model = lambda x: model(x)"""
-        else:
-            code = """fitted_model = lambda x: model(x, %s)""" % \
-               (', '.join([str(_) for _ in cof]))
-
-        # run the code
-        loc = dict(model=model)
-        exec(code, loc, loc)
-        return loc['fitted_model']
+        # Create a picklable function instead of lambda
+        def _create_fitted_model(model, cof):
+            if model.__name__ == "harmonize":
+                def fitted_model(x):
+                    return model(x)
+            else:
+                def fitted_model(x):
+                    return model(x, *cof)
+            return fitted_model
+        
+        return _create_fitted_model(model, cof)
 
     def _format_values_stack(self, values: np.ndarray) -> np.ndarray:
         """

--- a/test_kriging_parallelization.py
+++ b/test_kriging_parallelization.py
@@ -10,127 +10,127 @@ from skgstat import Variogram, OrdinaryKriging
 def test_serial_vs_parallel_consistency():
     """Test that serial and parallel execution give identical results."""
     print("Testing serial vs parallel consistency...")
-    
+
     # Create test data
     np.random.seed(42)
     coords = np.random.gamma(10, 4, size=(30, 2))
     values = np.random.normal(10, 2, size=30)
     V = Variogram(coords, values, model='gaussian', normalize=False)
-    
+
     # Serial execution
     ok_serial = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=1)
     with warnings.catch_warnings():
         warnings.simplefilter("ignore")  # Ignore kriging warnings for this test
         result_serial = ok_serial.transform(coords[:, 0], coords[:, 1])
-    
+
     # Parallel execution
     ok_parallel = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
     with warnings.catch_warnings():
         warnings.simplefilter("ignore")  # Ignore warnings for this test
         result_parallel = ok_parallel.transform(coords[:, 0], coords[:, 1])
-    
+
     # Compare results
     nan_mask_serial = np.isnan(result_serial)
     nan_mask_parallel = np.isnan(result_parallel)
-    
+
     # Check that NaN positions match
     if not np.array_equal(nan_mask_serial, nan_mask_parallel):
         print("  FAIL: NaN positions don't match between serial and parallel")
         return False
-    
+
     # Check that non-NaN values are close
     valid_mask = ~nan_mask_serial
     if not np.allclose(result_serial[valid_mask], result_parallel[valid_mask], rtol=1e-10):
         print("  FAIL: Results differ between serial and parallel execution")
         print(f"    Serial max diff: {np.max(np.abs(result_serial[valid_mask] - result_parallel[valid_mask]))}")
         return False
-    
+
     print("  PASS: Serial and parallel results are consistent")
     return True
 
 def test_different_n_jobs_values():
     """Test with different n_jobs values."""
     print("Testing different n_jobs values...")
-    
+
     np.random.seed(42)
     coords = np.random.gamma(10, 4, size=(20, 2))
     values = np.random.normal(10, 2, size=20)
     V = Variogram(coords, values, model='exponential', normalize=False)
-    
+
     n_jobs_values = [1, 2, 4]
     results = {}
-    
+
     for n_jobs in n_jobs_values:
         print(f"  Testing n_jobs={n_jobs}...")
         ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=n_jobs)
-        
+
         with warnings.catch_warnings():
             warnings.simplefilter("ignore")
             result = ok.transform(coords[:, 0], coords[:, 1])
-        
+
         results[n_jobs] = result
-    
+
     # Compare all results
     for i in range(len(n_jobs_values)):
         for j in range(i + 1, len(n_jobs_values)):
             n1, n2 = n_jobs_values[i], n_jobs_values[j]
             r1, r2 = results[n1], results[n2]
-            
+
             # Check NaN consistency
             nan_mask_1 = np.isnan(r1)
             nan_mask_2 = np.isnan(r2)
-            
+
             if not np.array_equal(nan_mask_1, nan_mask_2):
                 print(f"  FAIL: NaN positions differ between n_jobs={n1} and n_jobs={n2}")
                 return False
-            
+
             # Check value consistency
             valid_mask = ~nan_mask_1
             if not np.allclose(r1[valid_mask], r2[valid_mask], rtol=1e-10):
                 print(f"  FAIL: Results differ between n_jobs={n1} and n_jobs={n2}")
                 return False
-    
+
     print("  PASS: All n_jobs values give consistent results")
     return True
 
 def test_python_version_compatibility():
     """Test Python version specific behavior."""
     print("Testing Python version compatibility...")
-    
+
     import sys
     py_version = sys.version_info
-    
+
     print(f"  Python version: {py_version.major}.{py_version.minor}.{py_version.micro}")
-    
+
     if py_version >= (3, 13):
         print("  Testing Python 3.13+ warning behavior...")
-        
+
         np.random.seed(42)
         coords = np.random.gamma(10, 4, size=(15, 2))
         values = np.random.normal(10, 2, size=15)
         V = Variogram(coords, values, model='spherical', normalize=False)
-        
+
         with warnings.catch_warnings(record=True) as w:
             warnings.simplefilter("always")
             ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
             result = ok.transform(coords[:, 0], coords[:, 1])
-            
+
             # Check if warning was issued for Python 3.13+
             warning_found = any("multiprocessing" in str(warning.message).lower() for warning in w)
-            
+
             if warning_found:
                 print("  PASS: Appropriate warning issued for Python 3.13+")
             else:
                 print("  INFO: No multiprocessing warning (may be fixed)")
-    
+
     else:
         print("  Testing older Python version multiprocessing...")
-        
+
         np.random.seed(42)
         coords = np.random.gamma(10, 4, size=(15, 2))
         values = np.random.normal(10, 2, size=15)
         V = Variogram(coords, values, model='cubic', normalize=False)
-        
+
         try:
             ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
             result = ok.transform(coords[:, 0], coords[:, 1])
@@ -138,50 +138,50 @@ def test_python_version_compatibility():
         except Exception as e:
             print(f"  FAIL: Multiprocessing failed on Python {py_version.major}.{py_version.minor}: {e}")
             return False
-    
+
     return True
 
 def test_different_variogram_models():
     """Test with different variogram models."""
     print("Testing different variogram models...")
-    
+
     models = ['spherical', 'exponential', 'gaussian', 'cubic']
-    
+
     for model in models:
         print(f"  Testing {model} model...")
-        
+
         np.random.seed(42)
         coords = np.random.gamma(10, 4, size=(20, 2))
         values = np.random.normal(10, 2, size=20)
         V = Variogram(coords, values, model=model, normalize=False)
-        
+
         try:
             # Test serial
             ok_serial = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=1)
             with warnings.catch_warnings():
                 warnings.simplefilter("ignore")
                 result_serial = ok_serial.transform(coords[:, 0], coords[:, 1])
-            
+
             # Test parallel
             ok_parallel = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
             with warnings.catch_warnings():
                 warnings.simplefilter("ignore")
                 result_parallel = ok_parallel.transform(coords[:, 0], coords[:, 1])
-            
+
             # Check consistency
             nan_mask = np.isnan(result_serial) | np.isnan(result_parallel)
             valid_mask = ~nan_mask
-            
+
             if np.allclose(result_serial[valid_mask], result_parallel[valid_mask], rtol=1e-10):
                 print(f"    PASS: {model} model gives consistent results")
             else:
                 print(f"    FAIL: {model} model gives inconsistent results")
                 return False
-                
+
         except Exception as e:
             print(f"    FAIL: {model} model failed: {e}")
             return False
-    
+
     return True
 
 def run_all_tests():
@@ -189,17 +189,17 @@ def run_all_tests():
     print("=" * 60)
     print("COMPREHENSIVE KRIGING PARALLELIZATION TESTS")
     print("=" * 60)
-    
+
     tests = [
         test_serial_vs_parallel_consistency,
         test_different_n_jobs_values,
         test_python_version_compatibility,
         test_different_variogram_models,
     ]
-    
+
     passed = 0
     total = len(tests)
-    
+
     for test in tests:
         try:
             if test():
@@ -208,10 +208,10 @@ def run_all_tests():
         except Exception as e:
             print(f"  ERROR: Test failed with exception: {e}")
             print()
-    
+
     print("=" * 60)
     print(f"RESULTS: {passed}/{total} tests passed")
-    
+
     if passed == total:
         print("ALL TESTS PASSED! ✅")
         return True

--- a/test_kriging_parallelization.py
+++ b/test_kriging_parallelization.py
@@ -1,0 +1,224 @@
+#!/usr/bin/env python3
+"""
+Comprehensive test for OrdinaryKriging parallelization functionality.
+Tests Python 3.13+ compatibility and result consistency.
+"""
+import numpy as np
+import warnings
+from skgstat import Variogram, OrdinaryKriging
+
+def test_serial_vs_parallel_consistency():
+    """Test that serial and parallel execution give identical results."""
+    print("Testing serial vs parallel consistency...")
+    
+    # Create test data
+    np.random.seed(42)
+    coords = np.random.gamma(10, 4, size=(30, 2))
+    values = np.random.normal(10, 2, size=30)
+    V = Variogram(coords, values, model='gaussian', normalize=False)
+    
+    # Serial execution
+    ok_serial = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=1)
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")  # Ignore kriging warnings for this test
+        result_serial = ok_serial.transform(coords[:, 0], coords[:, 1])
+    
+    # Parallel execution
+    ok_parallel = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")  # Ignore warnings for this test
+        result_parallel = ok_parallel.transform(coords[:, 0], coords[:, 1])
+    
+    # Compare results
+    nan_mask_serial = np.isnan(result_serial)
+    nan_mask_parallel = np.isnan(result_parallel)
+    
+    # Check that NaN positions match
+    if not np.array_equal(nan_mask_serial, nan_mask_parallel):
+        print("  FAIL: NaN positions don't match between serial and parallel")
+        return False
+    
+    # Check that non-NaN values are close
+    valid_mask = ~nan_mask_serial
+    if not np.allclose(result_serial[valid_mask], result_parallel[valid_mask], rtol=1e-10):
+        print("  FAIL: Results differ between serial and parallel execution")
+        print(f"    Serial max diff: {np.max(np.abs(result_serial[valid_mask] - result_parallel[valid_mask]))}")
+        return False
+    
+    print("  PASS: Serial and parallel results are consistent")
+    return True
+
+def test_different_n_jobs_values():
+    """Test with different n_jobs values."""
+    print("Testing different n_jobs values...")
+    
+    np.random.seed(42)
+    coords = np.random.gamma(10, 4, size=(20, 2))
+    values = np.random.normal(10, 2, size=20)
+    V = Variogram(coords, values, model='exponential', normalize=False)
+    
+    n_jobs_values = [1, 2, 4]
+    results = {}
+    
+    for n_jobs in n_jobs_values:
+        print(f"  Testing n_jobs={n_jobs}...")
+        ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=n_jobs)
+        
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            result = ok.transform(coords[:, 0], coords[:, 1])
+        
+        results[n_jobs] = result
+    
+    # Compare all results
+    for i in range(len(n_jobs_values)):
+        for j in range(i + 1, len(n_jobs_values)):
+            n1, n2 = n_jobs_values[i], n_jobs_values[j]
+            r1, r2 = results[n1], results[n2]
+            
+            # Check NaN consistency
+            nan_mask_1 = np.isnan(r1)
+            nan_mask_2 = np.isnan(r2)
+            
+            if not np.array_equal(nan_mask_1, nan_mask_2):
+                print(f"  FAIL: NaN positions differ between n_jobs={n1} and n_jobs={n2}")
+                return False
+            
+            # Check value consistency
+            valid_mask = ~nan_mask_1
+            if not np.allclose(r1[valid_mask], r2[valid_mask], rtol=1e-10):
+                print(f"  FAIL: Results differ between n_jobs={n1} and n_jobs={n2}")
+                return False
+    
+    print("  PASS: All n_jobs values give consistent results")
+    return True
+
+def test_python_version_compatibility():
+    """Test Python version specific behavior."""
+    print("Testing Python version compatibility...")
+    
+    import sys
+    py_version = sys.version_info
+    
+    print(f"  Python version: {py_version.major}.{py_version.minor}.{py_version.micro}")
+    
+    if py_version >= (3, 13):
+        print("  Testing Python 3.13+ warning behavior...")
+        
+        np.random.seed(42)
+        coords = np.random.gamma(10, 4, size=(15, 2))
+        values = np.random.normal(10, 2, size=15)
+        V = Variogram(coords, values, model='spherical', normalize=False)
+        
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
+            result = ok.transform(coords[:, 0], coords[:, 1])
+            
+            # Check if warning was issued for Python 3.13+
+            warning_found = any("multiprocessing" in str(warning.message).lower() for warning in w)
+            
+            if warning_found:
+                print("  PASS: Appropriate warning issued for Python 3.13+")
+            else:
+                print("  INFO: No multiprocessing warning (may be fixed)")
+    
+    else:
+        print("  Testing older Python version multiprocessing...")
+        
+        np.random.seed(42)
+        coords = np.random.gamma(10, 4, size=(15, 2))
+        values = np.random.normal(10, 2, size=15)
+        V = Variogram(coords, values, model='cubic', normalize=False)
+        
+        try:
+            ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
+            result = ok.transform(coords[:, 0], coords[:, 1])
+            print("  PASS: Multiprocessing works on older Python versions")
+        except Exception as e:
+            print(f"  FAIL: Multiprocessing failed on Python {py_version.major}.{py_version.minor}: {e}")
+            return False
+    
+    return True
+
+def test_different_variogram_models():
+    """Test with different variogram models."""
+    print("Testing different variogram models...")
+    
+    models = ['spherical', 'exponential', 'gaussian', 'cubic']
+    
+    for model in models:
+        print(f"  Testing {model} model...")
+        
+        np.random.seed(42)
+        coords = np.random.gamma(10, 4, size=(20, 2))
+        values = np.random.normal(10, 2, size=20)
+        V = Variogram(coords, values, model=model, normalize=False)
+        
+        try:
+            # Test serial
+            ok_serial = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=1)
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                result_serial = ok_serial.transform(coords[:, 0], coords[:, 1])
+            
+            # Test parallel
+            ok_parallel = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
+            with warnings.catch_warnings():
+                warnings.simplefilter("ignore")
+                result_parallel = ok_parallel.transform(coords[:, 0], coords[:, 1])
+            
+            # Check consistency
+            nan_mask = np.isnan(result_serial) | np.isnan(result_parallel)
+            valid_mask = ~nan_mask
+            
+            if np.allclose(result_serial[valid_mask], result_parallel[valid_mask], rtol=1e-10):
+                print(f"    PASS: {model} model gives consistent results")
+            else:
+                print(f"    FAIL: {model} model gives inconsistent results")
+                return False
+                
+        except Exception as e:
+            print(f"    FAIL: {model} model failed: {e}")
+            return False
+    
+    return True
+
+def run_all_tests():
+    """Run all parallelization tests."""
+    print("=" * 60)
+    print("COMPREHENSIVE KRIGING PARALLELIZATION TESTS")
+    print("=" * 60)
+    
+    tests = [
+        test_serial_vs_parallel_consistency,
+        test_different_n_jobs_values,
+        test_python_version_compatibility,
+        test_different_variogram_models,
+    ]
+    
+    passed = 0
+    total = len(tests)
+    
+    for test in tests:
+        try:
+            if test():
+                passed += 1
+            print()
+        except Exception as e:
+            print(f"  ERROR: Test failed with exception: {e}")
+            print()
+    
+    print("=" * 60)
+    print(f"RESULTS: {passed}/{total} tests passed")
+    
+    if passed == total:
+        print("ALL TESTS PASSED! ✅")
+        return True
+    else:
+        print("SOME TESTS FAILED! ❌")
+        return False
+
+if __name__ == "__main__":
+    success = run_all_tests()
+    exit(0 if success else 1)

--- a/test_kriging_parallelization.py
+++ b/test_kriging_parallelization.py
@@ -188,7 +188,7 @@ def run_all_tests():
     tests = [
         test_serial_vs_parallel_consistency,
         test_different_n_jobs_values,
-        test_python_version_compatibility,
+        test_multiprocessing_works_across_python_versions,
         test_different_variogram_models,
     ]
 

--- a/test_kriging_parallelization.py
+++ b/test_kriging_parallelization.py
@@ -93,53 +93,48 @@ def test_different_n_jobs_values():
     print("  PASS: All n_jobs values give consistent results")
     return True
 
-def test_python_version_compatibility():
-    """Test Python version specific behavior."""
-    print("Testing Python version compatibility...")
+def test_multiprocessing_works_across_python_versions():
+    """Test that multiprocessing works on all supported Python versions."""
+    print("Testing multiprocessing across Python versions...")
 
     import sys
     py_version = sys.version_info
-
     print(f"  Python version: {py_version.major}.{py_version.minor}.{py_version.micro}")
 
-    if py_version >= (3, 13):
-        print("  Testing Python 3.13+ warning behavior...")
+    np.random.seed(42)
+    coords = np.random.gamma(10, 4, size=(20, 2))
+    values = np.random.normal(10, 2, size=20)
+    V = Variogram(coords, values, model='spherical', normalize=False)
 
-        np.random.seed(42)
-        coords = np.random.gamma(10, 4, size=(15, 2))
-        values = np.random.normal(10, 2, size=15)
-        V = Variogram(coords, values, model='spherical', normalize=False)
+    try:
+        # Test both serial and parallel execution
+        ok_serial = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=1)
+        ok_parallel = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
 
-        with warnings.catch_warnings(record=True) as w:
-            warnings.simplefilter("always")
-            ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
-            result = ok.transform(coords[:, 0], coords[:, 1])
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            result_serial = ok_serial.transform(coords[:, 0], coords[:, 1])
+            result_parallel = ok_parallel.transform(coords[:, 0], coords[:, 1])
 
-            # Check if warning was issued for Python 3.13+
-            warning_found = any("multiprocessing" in str(warning.message).lower() for warning in w)
+        # Verify results are consistent
+        nan_mask_serial = np.isnan(result_serial)
+        nan_mask_parallel = np.isnan(result_parallel)
 
-            if warning_found:
-                print("  PASS: Appropriate warning issued for Python 3.13+")
-            else:
-                print("  INFO: No multiprocessing warning (may be fixed)")
-
-    else:
-        print("  Testing older Python version multiprocessing...")
-
-        np.random.seed(42)
-        coords = np.random.gamma(10, 4, size=(15, 2))
-        values = np.random.normal(10, 2, size=15)
-        V = Variogram(coords, values, model='cubic', normalize=False)
-
-        try:
-            ok = OrdinaryKriging(V, min_points=5, max_points=15, n_jobs=2)
-            result = ok.transform(coords[:, 0], coords[:, 1])
-            print("  PASS: Multiprocessing works on older Python versions")
-        except Exception as e:
-            print(f"  FAIL: Multiprocessing failed on Python {py_version.major}.{py_version.minor}: {e}")
+        if not np.array_equal(nan_mask_serial, nan_mask_parallel):
+            print("  FAIL: NaN patterns differ between serial and parallel")
             return False
 
-    return True
+        valid_mask = ~nan_mask_serial
+        if not np.allclose(result_serial[valid_mask], result_parallel[valid_mask], rtol=1e-10):
+            print("  FAIL: Results differ between serial and parallel")
+            return False
+
+        print(f"  PASS: Multiprocessing works correctly on Python {py_version.major}.{py_version.minor}")
+        return True
+
+    except Exception as e:
+        print(f"  FAIL: Multiprocessing failed on Python {py_version.major}.{py_version.minor}: {e}")
+        return False
 
 def test_different_variogram_models():
     """Test with different variogram models."""

--- a/test_reproduce_issue.py
+++ b/test_reproduce_issue.py
@@ -9,29 +9,29 @@ def test_reproduce_error():
     """Reproduce the multiprocessing error from issue #201."""
     print("Testing OrdinaryKriging with n_jobs > 1 in Python 3.13...")
     print(f"Python version: {__import__('sys').version}")
-    
+
     # Create test data
     np.random.seed(42)
     coords = np.random.gamma(10, 4, size=(50, 2))
     values = np.random.normal(10, 2, size=50)
-    
+
     print(f"Created {len(coords)} test points")
-    
+
     # Create variogram
     V = Variogram(coords, values, model='gaussian', normalize=False)
     print(f"Created variogram: {V}")
-    
+
     try:
         # Test with n_jobs > 1 (should fail on Python 3.13)
         print("\n--- Testing with n_jobs=2 (parallel) ---")
         ok = OrdinaryKriging(V, min_points=5, max_points=20, n_jobs=2)
         print("OrdinaryKriging instance created successfully")
-        
+
         # This should trigger the error
         result = ok.transform(coords[:, 0], coords[:, 1])
         print(f"SUCCESS: Parallel kriging completed. Result shape: {result.shape}")
         return True
-        
+
     except Exception as e:
         print(f"ERROR: {type(e).__name__}: {e}")
         return False
@@ -39,15 +39,15 @@ def test_reproduce_error():
 def test_serial_baseline():
     """Test serial execution as baseline."""
     print("\n--- Testing with n_jobs=1 (serial baseline) ---")
-    
+
     # Create test data
     np.random.seed(42)
     coords = np.random.gamma(10, 4, size=(50, 2))
     values = np.random.normal(10, 2, size=50)
-    
+
     # Create variogram
     V = Variogram(coords, values, model='gaussian', normalize=False)
-    
+
     try:
         ok = OrdinaryKriging(V, min_points=5, max_points=20, n_jobs=1)
         result = ok.transform(coords[:, 0], coords[:, 1])
@@ -61,18 +61,18 @@ if __name__ == "__main__":
     print("=" * 60)
     print("Reproducing Python 3.13 multiprocessing issue #201")
     print("=" * 60)
-    
+
     # Test serial baseline
     serial_result = test_serial_baseline()
-    
+
     # Test parallel (should fail)
     parallel_success = test_reproduce_error()
-    
+
     print("\n" + "=" * 60)
     print("SUMMARY:")
     print(f"Serial execution: {'SUCCESS' if serial_result is not None else 'FAILED'}")
     print(f"Parallel execution: {'SUCCESS' if parallel_success else 'FAILED'}")
-    
+
     if not parallel_success:
         print("\nThis confirms the Python 3.13 multiprocessing issue.")
     else:

--- a/test_reproduce_issue.py
+++ b/test_reproduce_issue.py
@@ -5,9 +5,9 @@ Test script to reproduce the Python 3.13 multiprocessing error in OrdinaryKrigin
 import numpy as np
 from skgstat import Variogram, OrdinaryKriging
 
-def test_reproduce_error():
-    """Reproduce the multiprocessing error from issue #201."""
-    print("Testing OrdinaryKriging with n_jobs > 1 in Python 3.13...")
+def test_multiprocessing_fix():
+    """Test that the Python 3.13 multiprocessing fix works."""
+    print("Testing OrdinaryKriging multiprocessing fix...")
     print(f"Python version: {__import__('sys').version}")
 
     # Create test data
@@ -22,12 +22,12 @@ def test_reproduce_error():
     print(f"Created variogram: {V}")
 
     try:
-        # Test with n_jobs > 1 (should fail on Python 3.13)
+        # Test with n_jobs > 1 (should work with the fix)
         print("\n--- Testing with n_jobs=2 (parallel) ---")
         ok = OrdinaryKriging(V, min_points=5, max_points=20, n_jobs=2)
         print("OrdinaryKriging instance created successfully")
 
-        # This should trigger the error
+        # This should work with the fix
         result = ok.transform(coords[:, 0], coords[:, 1])
         print(f"SUCCESS: Parallel kriging completed. Result shape: {result.shape}")
         return True
@@ -59,21 +59,21 @@ def test_serial_baseline():
 
 if __name__ == "__main__":
     print("=" * 60)
-    print("Reproducing Python 3.13 multiprocessing issue #201")
+    print("Testing Python 3.13 multiprocessing fix for issue #201")
     print("=" * 60)
 
     # Test serial baseline
     serial_result = test_serial_baseline()
 
-    # Test parallel (should fail)
-    parallel_success = test_reproduce_error()
+    # Test parallel fix
+    parallel_success = test_multiprocessing_fix()
 
     print("\n" + "=" * 60)
     print("SUMMARY:")
     print(f"Serial execution: {'SUCCESS' if serial_result is not None else 'FAILED'}")
     print(f"Parallel execution: {'SUCCESS' if parallel_success else 'FAILED'}")
 
-    if not parallel_success:
-        print("\nThis confirms the Python 3.13 multiprocessing issue.")
+    if parallel_success:
+        print("\n✅ Python 3.13 multiprocessing fix is working!")
     else:
-        print("\nIssue appears to be resolved or not reproducible.")
+        print("\n❌ Python 3.13 multiprocessing fix needs more work.")

--- a/test_reproduce_issue.py
+++ b/test_reproduce_issue.py
@@ -1,0 +1,79 @@
+#!/usr/bin/env python3
+"""
+Test script to reproduce the Python 3.13 multiprocessing error in OrdinaryKriging.
+"""
+import numpy as np
+from skgstat import Variogram, OrdinaryKriging
+
+def test_reproduce_error():
+    """Reproduce the multiprocessing error from issue #201."""
+    print("Testing OrdinaryKriging with n_jobs > 1 in Python 3.13...")
+    print(f"Python version: {__import__('sys').version}")
+    
+    # Create test data
+    np.random.seed(42)
+    coords = np.random.gamma(10, 4, size=(50, 2))
+    values = np.random.normal(10, 2, size=50)
+    
+    print(f"Created {len(coords)} test points")
+    
+    # Create variogram
+    V = Variogram(coords, values, model='gaussian', normalize=False)
+    print(f"Created variogram: {V}")
+    
+    try:
+        # Test with n_jobs > 1 (should fail on Python 3.13)
+        print("\n--- Testing with n_jobs=2 (parallel) ---")
+        ok = OrdinaryKriging(V, min_points=5, max_points=20, n_jobs=2)
+        print("OrdinaryKriging instance created successfully")
+        
+        # This should trigger the error
+        result = ok.transform(coords[:, 0], coords[:, 1])
+        print(f"SUCCESS: Parallel kriging completed. Result shape: {result.shape}")
+        return True
+        
+    except Exception as e:
+        print(f"ERROR: {type(e).__name__}: {e}")
+        return False
+
+def test_serial_baseline():
+    """Test serial execution as baseline."""
+    print("\n--- Testing with n_jobs=1 (serial baseline) ---")
+    
+    # Create test data
+    np.random.seed(42)
+    coords = np.random.gamma(10, 4, size=(50, 2))
+    values = np.random.normal(10, 2, size=50)
+    
+    # Create variogram
+    V = Variogram(coords, values, model='gaussian', normalize=False)
+    
+    try:
+        ok = OrdinaryKriging(V, min_points=5, max_points=20, n_jobs=1)
+        result = ok.transform(coords[:, 0], coords[:, 1])
+        print(f"SUCCESS: Serial kriging completed. Result shape: {result.shape}")
+        return result
+    except Exception as e:
+        print(f"ERROR in serial execution: {type(e).__name__}: {e}")
+        return None
+
+if __name__ == "__main__":
+    print("=" * 60)
+    print("Reproducing Python 3.13 multiprocessing issue #201")
+    print("=" * 60)
+    
+    # Test serial baseline
+    serial_result = test_serial_baseline()
+    
+    # Test parallel (should fail)
+    parallel_success = test_reproduce_error()
+    
+    print("\n" + "=" * 60)
+    print("SUMMARY:")
+    print(f"Serial execution: {'SUCCESS' if serial_result is not None else 'FAILED'}")
+    print(f"Parallel execution: {'SUCCESS' if parallel_success else 'FAILED'}")
+    
+    if not parallel_success:
+        print("\nThis confirms the Python 3.13 multiprocessing issue.")
+    else:
+        print("\nIssue appears to be resolved or not reproducible.")

--- a/tutorials/README.rst
+++ b/tutorials/README.rst
@@ -15,4 +15,4 @@ Start with the basic tutorials to learn the fundamentals of SciKit-GStat:
 * **Tutorial 4**: Plotting - Visualizing variograms and results
 * **Tutorial 5**: Binning - Understanding and customizing distance binning
 * **Tutorial 6**: GSTools Integration - Using SciKit-GStat with GSTools
-* **Tutorial 7**: Maximum Likelihood - Fitting variograms using maximum likelihood estimation 
+* **Tutorial 7**: Maximum Likelihood - Fitting variograms using maximum likelihood estimation

